### PR TITLE
pg sources: harden schema change detection

### LINKED ISSUE #17927: pg sources: accepts schema change when columns have same name
### What version of Materialize are you using?

6dd84ed229cef7d65d11d605cadebe8bd3cadf8e

### How did you install Materialize?

Built from source

### What is the issue?

the following testdrive test passes on `main`

```
$ postgres-execute connection=postgres://postgres:postgres@postgres
ALTER USER postgres WITH replication;
DROP SCHEMA IF EXISTS public CASCADE;
DROP PUBLICATION IF EXISTS mz_source;

CREATE SCHEMA public;

CREATE TABLE alter_table_change_attnum (f1 VARCHAR(10), f2 VARCHAR(10));
ALTER TABLE alter_table_change_attnum REPLICA IDENTITY FULL;
INSERT INTO alter_table_change_attnum (f1, f2) VALUES ('f1_orig','f2_orig');

CREATE PUBLICATION mz_source FOR ALL TABLES;

> CREATE SOURCE mz_source
  FROM POSTGRES CONNECTION pg (PUBLICATION 'mz_source')
  FOR ALL TABLES;

> SELECT * from alter_table_rename_column;
f1_orig f2_orig

$ postgres-execute connection=postgres://postgres:postgres@postgres
ALTER TABLE alter_table_change_attnum DROP COLUMN f2;
ALTER TABLE alter_table_change_attnum ADD COLUMN f2 VARCHAR(10);
INSERT INTO alter_table_change_attnum (f1, f2) VALUES ('f1_changed', 'f2_changed');

# We have fooled the system into accepting this schema change while doing it online.
# Results should be
# f1_orig <null>
# f1_changed f2_changed
> SELECT * FROM alter_table_change_attnum;
f1_orig f2_orig
f1_changed f2_changed
```

However, this needs to error because the instance of `f2` in the final `SELECT` is not actually the same column.

My current thinking is that we can resolve this ambiguity with `attnum` which identifies the column with a monotonic number.

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by sploiselle**
We previously determined if table schemas changed based only on their columns' names--this is, of course, very easy to trick.

Because the PG replication stream provided a minimal amount of information, we have a somewhat limited number of options to resolve this issue:
- Error on any schema change
  - Pro: simple
  - Con: Precludes supporting adding/removing uningested columns because we will never be able to know if the columns in `Relation` messages are new columns with the old columns' names.
  
  The con here is too severe, so do not suggest we take this approach.
- Reach out to the PG source to determine the table's current schema
  - Pro: robust
  - Con: complex semantics, e.g. we have no way of tying together the `Relation` message in the replication stream to the table's current schema in the source. I have inlined comments about why this approach is sane.
  
  Notably, this approach also suggests removing the constraint information from `PgColumnDesc`, which was unused and confusing. If we do **not** remove this, it is most sane to error on changes to ingested columns' constraints, even though those constraints are not use anywhere within Materialize––seems like an unnecessarily punitive approach.

### Motivation

This PR fixes a recognized bug. Fixes #17927

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

**Comment by sploiselle** [comment]
Whoops--this needs to upgrade the serialized PG column defs; moving back to draft temporarily.

**Comment by sploiselle** [comment]
Ok--good to review

File: src/adapter/src/catalog/migrate.rs
Comment (sploiselle) [inline src/adapter/src/catalog/migrate.rs:11]: Will fix import order

**Comment by philip-stoev** [comment]
Please give me until tomorrow to ponder on this.

File: src/adapter/src/catalog/migrate.rs
Comment (jkosh44) [inline src/adapter/src/catalog/migrate.rs:32]: Maybe we should have two variants of `rewrite_items`, one with a `ConnCatalog` and one without. That way we don't have have this `Option` and `expect` below.

Alternatively, if we implement `Clone` for `ConnCatalog`, we don't need the `Option` and can just clone the `ConnCatalog` before passing it into the closure when we call this function.

File: src/adapter/src/catalog/migrate.rs
Comment (jkosh44) [inline src/adapter/src/catalog/migrate.rs:227]: Is it 0 here because 0 is the default value? I think in general we're trying to add an `Option` to new fields and then remove the `Option` in the next release.

File: src/adapter/src/catalog/migrate.rs
Comment (philip-stoev) [inline src/adapter/src/catalog/migrate.rs:272]: Should this situation the status of the source to be updated in mz_source_statuses? Right now it will be just a `WARN` message in the logs, right?

What would be an "external dependency" in this case? Should the reason for the error be part of the message rather than consumed and discarded via `_` ?

File: src/adapter/src/catalog/migrate.rs
Comment (philip-stoev) [inline src/adapter/src/catalog/migrate.rs:275]: restart of the source or of Mz?

File: src/adapter/src/catalog/migrate.rs
Comment (philip-stoev) [reply src/adapter/src/catalog/migrate.rs:272]: Nevermind, the error does end up in `mz_source_statuses` , so all good.

Comment (philip-stoev) [reply src/adapter/src/catalog/migrate.rs:272]: The error in the mz_source_statuses is `publication "mz_source" does not exist` . The longer warning is only in the log, so not actionable by users.

**Review by philip-stoev — APPROVED** [review]
I built a new dedicated test that is a superset of the pg-cdc and pg-cdc-resumption mzcompose tests in the sense that it performs a bunch of Pg-side schema changes in different contexts (while Mz is down, while Mz is up, while Mz is being upgraded, if Postgres is down while Mz restarts, etc.).

There were no correctness issues, but there were various interesting things that may or may not be specific to your branch. I suggest you merge your branch as it is and then I will file those as follow-up tickets and push the test as a separate PR.

1. `DROP PUBLICATION` may or may not be cause the source to error out -- it may simply stop receiving new data without any other indications
2. `DROP PUBLICATION` may or may not cause the source to **partially error-out**: an error is reflected in the `mz_source_statuses` table, but selecting from the source does not return an error - it continues to return the data from before the publication was dropped.
3. The error at the source for the same situation on the Pg side may be either `Schema for table t1 differs` or `table has been altered`, depending on one's luck. Does this imply that there are two separate code paths that decide on whether the two schemas are compatible or not? If so, maybe those two paths may reach diverging decisions for the same situation? 
4. `ADD COLUMN` will either cause the source to error out (in two separate ways nonetheless, either with `table has been altered` or `extra data after last expected column`) or be ignored -- the source may continue to ingest new rows and outputs the values from the previously-existing columns.

File: src/postgres-util/src/desc.rs
Comment (sploiselle) [reply src/postgres-util/src/desc.rs:93]: See [727a837](https://github.com/MaterializeInc/materialize/pull/17987/commits/727a837a2e0183c200d4b44ef35c17743baf86b7)

File: src/adapter/src/catalog/migrate.rs
Comment (sploiselle) [reply src/adapter/src/catalog/migrate.rs:32]: The migration code thrashes a lot, so I wanted to localize these changes as much as possible. I think that the aesthetics of doing something more robust are desirable but this code will be yanked out in two weeks so didn't want to spend any more energy on it than was necessary.

Comment (jkosh44) [reply src/adapter/src/catalog/migrate.rs:32]: Yeah good point, we have an issue open somewhere to re-work all of this anyways.

File: src/adapter/src/catalog/migrate.rs
Comment (sploiselle) [reply src/adapter/src/catalog/migrate.rs]: Fixed

File: src/adapter/src/catalog/migrate.rs
Comment (sploiselle) [reply src/adapter/src/catalog/migrate.rs:272]: Yes, I think that behavior is right. I don't want the migration to prevent Materialize from booting and whatever the issue is will get surfaced elsewhere. The warnings here are just for our debugging purposes.

File: src/adapter/src/catalog/migrate.rs
Comment (sploiselle) [reply src/adapter/src/catalog/migrate.rs]: Sure thing--did this.

**Comment by sploiselle** [comment]
@petrosagg @aljoscha Do one of you mind reviewing from a storage perspective? ty

File: src/postgres-util/src/desc.rs
Comment (aljoscha) [inline src/postgres-util/src/desc.rs:29]: on order of their column ID?

**Review by aljoscha — COMMENTED** [review]
This look good! But I'd feel better if Petros did a final review since he knows the PG source better than I do.

I'm not sure if the back-and-forth on the Protobuf fields for the column desc (the later commit, that removes the nullable/primary_key constraints)  is needed, but I'll leave that up to you.

**Comment by sploiselle** [comment]
@petrosagg Can I get a final review of this?

File: src/postgres-util/src/desc.proto
Comment (petrosagg) [inline src/postgres-util/src/desc.proto:27]: I think it'd be a good idea to leave those in even if we don't currently use them because it will avoid us from having to do another migration in the future and we want to use them anyway. Especially the nullability information should be very easy to incorporate when we define the `RelationDesc` of the subsource in `pure.rs`

File: src/storage/src/source/postgres.rs
Comment (sploiselle) [reply src/storage/src/source/postgres.rs:1170]: Yes, the repro is in #17927; it seems that PG consolidates DDL changes because we never detect that the schema changed.

File: src/postgres-util/src/desc.proto
Comment (sploiselle) [reply src/postgres-util/src/desc.proto]: I fixed this to keep and use the nullability information. The key information is confusing because it both isn't used and isn't useful on its own. If we ever want to support the notion of keys, we'll already have to write a migration to determine which columns are in the keys, so we can repopulate this then.

File: src/storage/src/source/postgres.rs
Comment (sploiselle) [reply src/storage/src/source/postgres.rs:1171]: This looks to be the same to me; let me know if I'm losing my mind.

File: src/storage/src/source/postgres.rs
Comment (sploiselle) [reply src/storage/src/source/postgres.rs]: Oh yes--fixed

File: src/postgres-util/src/desc.proto
Comment (sploiselle) [reply src/postgres-util/src/desc.proto]: Chatted with Petros about the primary key information and agreed to move the key information onto the table description.

**Comment by sploiselle** [comment]
Converting back to draft--realized that adding constraint support introduces constraints into existing PG tables' definitions and that then means that we need to express compatibility between our recorded version of the schema and the upstream schema.

**Comment by sploiselle** [comment]
This was done in #18313 + #18403

**Pull request closed without merging**
